source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp

```diff
@@ -0,0 +1,637 @@
+
+      m_filepath       = rsl::exchange(other.m_filepath, "");
+      m_queued_request = rsl::exchange(other.m_queued_request, nullptr);
+      m_is_done        = rsl::exchange(other.m_is_done, false);
+      m_buffer         = rsl::exchange(other.m_buffer, nullptr);
+      m_size           = rsl::exchange(other.m_size, 0_bytes);
+
+      m_queued_request->swap_request_to_signal(&other, this);
+
+      return *this;
+    }
+
+    void ReadRequest::signal(const rsl::byte* buffer, rsl::memory_size size)
+    {
+      m_is_done = true;
+      m_buffer  = buffer;
+      m_size    = size;
+    }
+
+    void ReadRequest::wait() const
+    {
+      // this should ideally be solved with fibers..
+      while(!m_is_done)
+      {
+        using namespace rsl::chrono_literals; // NOLINT(google-build-using-namespace)
+        rsl::this_thread::sleep_for(1ms);
+      }
+    }
+
+    const rsl::byte* ReadRequest::data() const
+    {
+      return m_buffer;
+    }
+    rsl::memory_size ReadRequest::count() const
+    {
+      return m_size;
+    }
+
+    rsl::string_view ReadRequest::filepath() const
+    {
+      return m_filepath;
+    }
+
+    // NOLINTBEGIN(cppcoreguidelines-avoid-non-const-global-variables, fuchsia-statically-constructed-objects)
+
+    // This the root where all relative paths will start from
+    rsl::medium_stack_string g_root;
+
+    // This controls the state of the vfs
+    StateController<VfsState> g_vfs_state_controller(VfsState::NotInitialized);
+    
+    // mutices for the asyncronous operation of the vfs
+    rsl::mutex g_read_request_mutex;
+    rsl::mutex g_closed_request_mutex;
+
+    // queues the vfs uses for its async operations
+    rsl::vector<rsl::string_view> g_read_requests_in_order;
+    rsl::unordered_map<rsl::string, rsl::unique_ptr<QueuedRequest>> g_read_requests;
+    rsl::vector<rsl::unique_ptr<QueuedRequest>> g_closed_requests;
+
+    // mounted roots
+    rsl::unordered_map<MountingPoint, rsl::string> g_mounted_roots;
+
+    // threads used by the vfs to perform the async operations
+    rsl::thread g_reading_thread;
+    rsl::thread g_closing_thread;
+
+    // NOLINTEND(cppcoreguidelines-avoid-non-const-global-variables, fuchsia-statically-constructed-objects)
+
+    void process_read_requests()
+    {
+      while (g_vfs_state_controller.has_state(VfsState::Running))
+      {
+        rsl::unique_lock lock(g_read_request_mutex);
+        if (!g_read_requests.empty())
+        {
+          // get the queued request and remvoe it from the queue,
+          // but remove in the reverse order that they got added
+          rsl::string_view filepath = g_read_requests_in_order.front();
+          g_read_requests_in_order.erase(g_read_requests_in_order.cbegin());
+          rsl::unique_ptr<QueuedRequest> request = rsl::move(g_read_requests[filepath]);
+          g_read_requests.erase(filepath);
+
+          // we don't need access to the queue anymore, we can unlock its access mutex
+          lock.unlock();
```

Comment:
NIT: I always add another scope to make sure the dtor is triggered, if this code ever changes it's done automatically without me thinking about the unlock.

Reply:
You added a thump here but I don't see this change? 
Was this just to verify you read it?

Reply:
I put an extra scope around the entire if block, it's impossible for me to start before the if and end halfway through